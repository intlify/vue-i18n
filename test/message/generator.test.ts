/* eslint-disable no-irregular-whitespace */

import { createParser } from '../../src/message/parser'
import { transform } from '../../src/message/transformer'
import { generate } from '../../src/message/generator'

describe('text', () => {
  test('basic', () => {
    const parser = createParser()
    const msg = 'hello world'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(`"hello world"`)
    expect(code).toMatch(`])`)
  })

  test('multline', () => {
    const parser = createParser()
    const msg = 'hello\n world'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(`"hello\\n world"`)
    expect(code).toMatch(`])`)
  })
})

describe('list', () => {
  test('basic', () => {
    const parser = createParser()
    const msg = 'hi {0} !'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(`"hi ", _interpolate(_list(0)), " !"`)
    expect(code).toMatch(`])`)
  })

  test('multiple', () => {
    const parser = createParser()
    const msg = '{0} {1} !'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(
      `_interpolate(_list(0)), " ", _interpolate(_list(1)), " !"`
    )
    expect(code).toMatch(`])`)
  })
})

describe('named', () => {
  test('basic', () => {
    const parser = createParser()
    const msg = 'hi {name} !'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(`"hi ", _interpolate(_named("name")), " !"`)
    expect(code).toMatch(`])`)
  })

  test('multiple', () => {
    const parser = createParser()
    const msg = '{greeting} {name} !'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(
      `_interpolate(_named("greeting")), " ", _interpolate(_named("name")), " !"`
    )
    expect(code).toMatch(`])`)
  })
})

describe('literal', () => {
  test('ascii', () => {
    const parser = createParser()
    const msg = `hi {'kazupon'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('multibytes', () => {
    const parser = createParser()
    const msg = `hi {'かずぽん'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('emoji', () => {
    const parser = createParser()
    const msg = `hi {'😺'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('{}', () => {
    const parser = createParser()
    const msg = `{'{}'}`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('!#%^&*()-_+=[]:;?.<>"`', () => {
    const parser = createParser()
    const msg = `hi {'${'!#%^&*()-_+=[]:;?.<>"`'}'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('escaped single quote', () => {
    const parser = createParser()
    const msg = `hi {'\\''} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('escaped slash', () => {
    const parser = createParser()
    const msg = `hi {'\\\\'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('unicode 4 digits', () => {
    const parser = createParser()
    const msg = `hi {'${'\u0041'}'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('escaped unicode 4 digits', () => {
    const parser = createParser()
    const msg = `hi {'\\\\u0041'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('unicode 6 digits', () => {
    const parser = createParser()
    const msg = `hi {'${'U01F602'}'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })

  test('escaped unicode 6 digits', () => {
    const parser = createParser()
    const msg = `hi {'\\\\U01F602'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
  })
})

describe('linked', () => {
  test('key', () => {
    const parser = createParser()
    const msg = 'hi @:name !'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(`"hi ", _message("name")(ctx), " !"`)
    expect(code).toMatch(`])`)
  })

  test('list', () => {
    const parser = createParser()
    const msg = 'hi @:{0} !'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(`"hi ", _message(_interpolate(_list(0)))(ctx), " !"`)
    expect(code).toMatch(`])`)
  })

  test('named', () => {
    const parser = createParser()
    const msg = 'hi @:{name} !'
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(
      `"hi ", _message(_interpolate(_named("name")))(ctx), " !"`
    )
    expect(code).toMatch(`])`)
  })

  test('modifier', () => {
    const parser = createParser()
    const msg = `hi @.upper:{'name'} !`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return _normalize([`)
    expect(code).toMatch(
      `"hi ", _modifier("upper")(_message("name")(ctx), _type), " !"`
    )
    expect(code).toMatch(`])`)
  })
})

describe('plural', () => {
  test('simple', () => {
    const parser = createParser()
    const msg = 'no apples | one apple  |  too much apples  '
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`_normalize([`)
    expect(code).toMatch(`"no apples"`)
    expect(code).toMatch(` ]), _normalize([`)
    expect(code).toMatch(`"one apple"`)
    expect(code).toMatch(` ]), _normalize([`)
    expect(code).toMatch(`"too much apples  "`)
    expect(code).toMatch(` ])`)
    expect(code).toMatch(`][_pluralRule(_pluralIndex, 3, _orgPluralRule)]`)
  })

  test('complex', () => {
    const parser = createParser()
    const msg = `@.caml:{'no apples'} | {0} apple | {n}　apples`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast)
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`_normalize([`)
    expect(code).toMatch(`_modifier("caml")(_message("no apples")(ctx), _type)`)
    expect(code).toMatch(` ]), _normalize([`)
    expect(code).toMatch(`_interpolate(_list(0)), " apple"`)
    expect(code).toMatch(` ]), _normalize([`)
    expect(code).toMatch(`_interpolate(_named("n")), "　apples"`)
    expect(code).toMatch(` ])`)
    expect(code).toMatch(`][_pluralRule(_pluralIndex, 3, _orgPluralRule)]`)
  })
})

describe('mode', () => {
  test('arrow', () => {
    const parser = createParser()
    const msg = `@.caml:{'no apples'} | {0} apple | {n}　apples`
    const ast = parser.parse(msg)
    transform(ast)
    const code = generate(ast, { mode: 'arrow' })
    expect(code).toMatchSnapshot(msg)
    expect(code).toMatch(`(ctx) => {`)
  })
})

/* eslint-enable no-irregular-whitespace */
