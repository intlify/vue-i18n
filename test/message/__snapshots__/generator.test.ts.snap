// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linked key 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked key 2`] = `
Object {
  "generatedColumn": 20,
  "generatedLine": 4,
  "name": "name",
  "originalColumn": 5,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked key 3`] = `
Object {
  "generatedColumn": 34,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 9,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked key: hi @:name ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, message: _message } = ctx
  return _normalize([
    \\"hi \\", _message(\\"name\\")(ctx), \\" !\\"
  ])
}"
`;

exports[`linked list 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked list 2`] = `
Object {
  "generatedColumn": 20,
  "generatedLine": 4,
  "name": "0",
  "originalColumn": 5,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked list 3`] = `
Object {
  "generatedColumn": 50,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 8,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked list: hi @:{0} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, list: _list, message: _message } = ctx
  return _normalize([
    \\"hi \\", _message(_interpolate(_list(0)))(ctx), \\" !\\"
  ])
}"
`;

exports[`linked modifier 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked modifier 2`] = `
Object {
  "generatedColumn": 21,
  "generatedLine": 4,
  "name": "upper",
  "originalColumn": 4,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked modifier 3`] = `
Object {
  "generatedColumn": 39,
  "generatedLine": 4,
  "name": "name",
  "originalColumn": 11,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked modifier 4`] = `
Object {
  "generatedColumn": 61,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 19,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked modifier: hi @.upper:{'name'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, modifier: _modifier, type: _type, message: _message } = ctx
  return _normalize([
    \\"hi \\", _modifier(\\"upper\\")(_message(\\"name\\")(ctx), _type), \\" !\\"
  ])
}"
`;

exports[`linked named 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked named 2`] = `
Object {
  "generatedColumn": 20,
  "generatedLine": 4,
  "name": "name",
  "originalColumn": 5,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked named 3`] = `
Object {
  "generatedColumn": 56,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 11,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`linked named: hi @:{name} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, named: _named, message: _message } = ctx
  return _normalize([
    \\"hi \\", _message(_interpolate(_named(\\"name\\")))(ctx), \\" !\\"
  ])
}"
`;

exports[`list basic 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`list basic 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "0",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`list basic 3`] = `
Object {
  "generatedColumn": 35,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 6,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`list basic: hi {0} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, list: _list } = ctx
  return _normalize([
    \\"hi \\", _interpolate(_list(0)), \\" !\\"
  ])
}"
`;

exports[`list multiple 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "0",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`list multiple 2`] = `
Object {
  "generatedColumn": 28,
  "generatedLine": 4,
  "name": " ",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`list multiple 3`] = `
Object {
  "generatedColumn": 33,
  "generatedLine": 4,
  "name": "1",
  "originalColumn": 4,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`list multiple 4`] = `
Object {
  "generatedColumn": 57,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 7,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`list multiple: {0} {1} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, list: _list } = ctx
  return _normalize([
    _interpolate(_list(0)), \\" \\", _interpolate(_list(1)), \\" !\\"
  ])
}"
`;

exports[`literal !#%^&*()-_+=[]:;?.<>"\` 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal !#%^&*()-_+=[]:;?.<>"\` 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "!#%^&*()-_+=[]:;?.<>\\"\`",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal !#%^&*()-_+=[]:;?.<>"\` 3`] = `
Object {
  "generatedColumn": 38,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 29,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal !#%^&*()-_+=[]:;?.<>"\`: hi {'!#%^&*()-_+=[]:;?.<>"\`'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"!#%^&*()-_+=[]:;?.<>\\\\\\"\`\\", \\" !\\"
  ])
}"
`;

exports[`literal {} 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "{}",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal {}: {'{}'} 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"{}\\"
  ])
}"
`;

exports[`literal ascii 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal ascii 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "kazupon",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal ascii 3`] = `
Object {
  "generatedColumn": 22,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 14,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal ascii: hi {'kazupon'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"kazupon\\", \\" !\\"
  ])
}"
`;

exports[`literal emoji 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal emoji 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "😺",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal emoji 3`] = `
Object {
  "generatedColumn": 17,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 9,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal emoji: hi {'😺'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"😺\\", \\" !\\"
  ])
}"
`;

exports[`literal escaped single quote 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped single quote 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "'",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped single quote 3`] = `
Object {
  "generatedColumn": 16,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 9,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped single quote: hi {'\\''} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"'\\", \\" !\\"
  ])
}"
`;

exports[`literal escaped slash 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped slash 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "\\\\",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped slash 3`] = `
Object {
  "generatedColumn": 17,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 9,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped slash: hi {'\\\\'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"\\\\\\\\\\", \\" !\\"
  ])
}"
`;

exports[`literal escaped unicode 4 digits 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped unicode 4 digits 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "\\\\u0041",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped unicode 4 digits 3`] = `
Object {
  "generatedColumn": 22,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 14,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped unicode 4 digits: hi {'\\\\u0041'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"\\\\\\\\u0041\\", \\" !\\"
  ])
}"
`;

exports[`literal escaped unicode 6 digits 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped unicode 6 digits 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "\\\\U01F602",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped unicode 6 digits 3`] = `
Object {
  "generatedColumn": 24,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 16,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal escaped unicode 6 digits: hi {'\\\\U01F602'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"\\\\\\\\U01F602\\", \\" !\\"
  ])
}"
`;

exports[`literal multibytes 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal multibytes 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "かずぽん",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal multibytes 3`] = `
Object {
  "generatedColumn": 19,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 11,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal multibytes: hi {'かずぽん'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"かずぽん\\", \\" !\\"
  ])
}"
`;

exports[`literal unicode 4 digits 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal unicode 4 digits 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "A",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal unicode 4 digits 3`] = `
Object {
  "generatedColumn": 16,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 8,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal unicode 4 digits: hi {'A'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"A\\", \\" !\\"
  ])
}"
`;

exports[`literal unicode 6 digits 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal unicode 6 digits 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "U01F602",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal unicode 6 digits 3`] = `
Object {
  "generatedColumn": 22,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 14,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`literal unicode 6 digits: hi {'U01F602'} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hi \\", \\"U01F602\\", \\" !\\"
  ])
}"
`;

exports[`mode arrow 1`] = `
Object {
  "generatedColumn": 16,
  "generatedLine": 5,
  "name": "caml",
  "originalColumn": 1,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`mode arrow 2`] = `
Object {
  "generatedColumn": 33,
  "generatedLine": 5,
  "name": "no apples",
  "originalColumn": 7,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`mode arrow 3`] = `
Object {
  "generatedColumn": 6,
  "generatedLine": 7,
  "name": "0",
  "originalColumn": 23,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`mode arrow 4`] = `
Object {
  "generatedColumn": 30,
  "generatedLine": 7,
  "name": " apple",
  "originalColumn": 26,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`mode arrow 5`] = `
Object {
  "generatedColumn": 6,
  "generatedLine": 9,
  "name": "n",
  "originalColumn": 35,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`mode arrow 6`] = `
Object {
  "generatedColumn": 33,
  "generatedLine": 9,
  "name": "　apples",
  "originalColumn": 38,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`mode arrow: @.caml:{'no apples'} | {0} apple | {n}　apples 1`] = `
"(ctx) => {
  const { normalize: _normalize, modifier: _modifier, type: _type, message: _message, interpolate: _interpolate, list: _list, named: _named, pluralIndex: _pluralIndex, pluralRule: _pluralRule, orgPluralRule: _orgPluralRule } = ctx
  return [
    _normalize([
      _modifier(\\"caml\\")(_message(\\"no apples\\")(ctx), _type)
    ]), _normalize([
      _interpolate(_list(0)), \\" apple\\"
    ]), _normalize([
      _interpolate(_named(\\"n\\")), \\"　apples\\"
    ])
  ][_pluralRule(_pluralIndex, 3, _orgPluralRule)]
}"
`;

exports[`named basic 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hi ",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`named basic 2`] = `
Object {
  "generatedColumn": 11,
  "generatedLine": 4,
  "name": "name",
  "originalColumn": 3,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`named basic 3`] = `
Object {
  "generatedColumn": 41,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 9,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`named basic: hi {name} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, named: _named } = ctx
  return _normalize([
    \\"hi \\", _interpolate(_named(\\"name\\")), \\" !\\"
  ])
}"
`;

exports[`named multiple 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "greeting",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`named multiple 2`] = `
Object {
  "generatedColumn": 38,
  "generatedLine": 4,
  "name": " ",
  "originalColumn": 10,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`named multiple 3`] = `
Object {
  "generatedColumn": 43,
  "generatedLine": 4,
  "name": "name",
  "originalColumn": 11,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`named multiple 4`] = `
Object {
  "generatedColumn": 73,
  "generatedLine": 4,
  "name": " !",
  "originalColumn": 17,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`named multiple: {greeting} {name} ! 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, interpolate: _interpolate, named: _named } = ctx
  return _normalize([
    _interpolate(_named(\\"greeting\\")), \\" \\", _interpolate(_named(\\"name\\")), \\" !\\"
  ])
}"
`;

exports[`plural complex 1`] = `
Object {
  "generatedColumn": 16,
  "generatedLine": 5,
  "name": "caml",
  "originalColumn": 1,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural complex 2`] = `
Object {
  "generatedColumn": 33,
  "generatedLine": 5,
  "name": "no apples",
  "originalColumn": 7,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural complex 3`] = `
Object {
  "generatedColumn": 6,
  "generatedLine": 7,
  "name": "0",
  "originalColumn": 23,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural complex 4`] = `
Object {
  "generatedColumn": 30,
  "generatedLine": 7,
  "name": " apple",
  "originalColumn": 26,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural complex 5`] = `
Object {
  "generatedColumn": 6,
  "generatedLine": 9,
  "name": "n",
  "originalColumn": 35,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural complex 6`] = `
Object {
  "generatedColumn": 33,
  "generatedLine": 9,
  "name": "　apples",
  "originalColumn": 38,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural complex: @.caml:{'no apples'} | {0} apple | {n}　apples 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, modifier: _modifier, type: _type, message: _message, interpolate: _interpolate, list: _list, named: _named, pluralIndex: _pluralIndex, pluralRule: _pluralRule, orgPluralRule: _orgPluralRule } = ctx
  return [
    _normalize([
      _modifier(\\"caml\\")(_message(\\"no apples\\")(ctx), _type)
    ]), _normalize([
      _interpolate(_list(0)), \\" apple\\"
    ]), _normalize([
      _interpolate(_named(\\"n\\")), \\"　apples\\"
    ])
  ][_pluralRule(_pluralIndex, 3, _orgPluralRule)]
}"
`;

exports[`plural simple 1`] = `
Object {
  "generatedColumn": 6,
  "generatedLine": 5,
  "name": "no apples",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural simple 2`] = `
Object {
  "generatedColumn": 6,
  "generatedLine": 7,
  "name": "one apple",
  "originalColumn": 12,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural simple 3`] = `
Object {
  "generatedColumn": 6,
  "generatedLine": 9,
  "name": "too much apples  ",
  "originalColumn": 26,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`plural simple: no apples | one apple  |  too much apples   1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize, pluralIndex: _pluralIndex, pluralRule: _pluralRule, orgPluralRule: _orgPluralRule } = ctx
  return [
    _normalize([
      \\"no apples\\"
    ]), _normalize([
      \\"one apple\\"
    ]), _normalize([
      \\"too much apples  \\"
    ])
  ][_pluralRule(_pluralIndex, 3, _orgPluralRule)]
}"
`;

exports[`text basic 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hello world",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "message.intl",
}
`;

exports[`text basic: hello world 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hello world\\"
  ])
}"
`;

exports[`text multline 1`] = `
Object {
  "generatedColumn": 4,
  "generatedLine": 4,
  "name": "hello
 world",
  "originalColumn": 0,
  "originalLine": 1,
  "source": "foo.bar",
}
`;

exports[`text multline: hello
 world 1`] = `
"function __msg__ (ctx) {
  const { normalize: _normalize } = ctx
  return _normalize([
    \\"hello\\\\n world\\"
  ])
}"
`;
